Main.hs

```diff
@@ -4,26 +4,100 @@ import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = NumExpr Double
+  | SqrtExpr Expr
+  | AddExpr Expr Expr
+  | SubExpr Expr Expr
+  | MulExpr Expr Expr
+  | DivExpr Expr Expr
+  | PowerExpr Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (NumExpr a) = show a
+  show (SqrtExpr a) = "(" ++ show a ++ ")"
+  show (AddExpr a b) = "(" ++ show a ++ "+" ++ show b ++ ")"
+  show (SubExpr a b) = "(" ++ show a ++ "-" ++ show b ++ ")"
+  show (MulExpr a b) = "(" ++ show a ++ "*" ++ show b ++ ")"
+  show (DivExpr a b) = "(" ++ show a ++ "/" ++ show b ++ ")"
+  show (PowerExpr a b) = "(" ++ show a ++ "^" ++ show b ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  NumExpr a == NumExpr b = a == b
+  SqrtExpr a == SqrtExpr b = a == b
+  AddExpr a b == AddExpr c d = (a==c) && (b==d)
+  SubExpr a b == SubExpr c d = (a==c) && (b==d)
+  MulExpr a b == MulExpr c d = (a==c) && (b==d)
+  DivExpr a b == DivExpr c d = (a==c) && (b==d)
+  PowerExpr a b == PowerExpr c d = (a==c) && (b==d)
+  _ == _ = False
 
 data Error
+  = DivByZero
+  | NegativeSqrtNum
```

Comment:
It would be helpful if an error contained info about the sub-expression, which produced it.

Reply:
I think it's fixed now, I added some additional info for each sub-expression.